### PR TITLE
Add Lognormal and Weibull distributions

### PR DESCRIPTION
**Pull request by infinity0**

### CODE REVIEW

File: Statistics/Distribution/Weibull.hs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    ++ show l
+
+logDensity :: WeibullDistribution -> Double -> Double
+logDensity (WD k l) x
+  | x < 0     = 0
+  | otherwise = log k + (k - 1) * log x - k * log l - (x / l) ** k
+
+cumulative :: WeibullDistribution -> Double -> Double
+cumulative (WD k l) x | x < 0     = 0
+                      | otherwise = 1 - exp (-(x / l) ** k)
```
Comment (Shimuuar) [inline Statistics/Distribution/Weibull.hs:130]: Please use [expm1](https://hackage.haskell.org/package/math-functions-0.3.4.0/docs/Numeric-SpecFunctions.html#v:expm1). `e^x - 1` is well known sourse of precision loss for small x

File: Statistics/Distribution/Weibull.hs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+complCumulative :: WeibullDistribution -> Double -> Double
+complCumulative (WD k l) x | x < 0     = 0
+                           | otherwise = exp (-(x / l) ** k)
+
+quantile :: WeibullDistribution -> Double -> Double
+quantile (WD k l) p
+  | p == 0         = 0
+  | p == 1         = inf
+  | p > 0 && p < 1 = l * (-log (1 - p)) ** (1 / k)
```
Comment (Shimuuar) [inline Statistics/Distribution/Weibull.hs:140]: Simlarly please use log1p

File: Statistics/Distribution/Weibull.hs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+instance D.ContDistr WeibullDistribution where
+  logDensity    = logDensity
+  quantile      = quantile
+  complQuantile = complQuantile
+
+instance D.MaybeMean WeibullDistribution where
+  maybeMean = Just . D.mean
+
+instance D.Mean WeibullDistribution where
+  mean (WD k l) = l * exp (logGamma (1 + 1 / k))
```
Comment (Shimuuar) [inline Statistics/Distribution/Weibull.hs:67]: This whole expression just screams: "I lose precision for large k". I however not sure whether it's possible to rewrite it so it won't without adding more functions to `math-functions`

However it's even worse for variance and probably even impossible to fix

File: Statistics/Distribution/Lognormal.hs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+   where
+    m = D.mean d
+    v = D.variance d
+
+instance D.MaybeVariance LognormalDistribution where
+  maybeStdDev   = Just . D.stdDev
+  maybeVariance = Just . D.variance
+
+instance D.Variance LognormalDistribution where
+  variance (LND d) = (exp v - 1) * exp (2 * m + v)
```
Comment (Shimuuar) [inline Statistics/Distribution/Lognormal.hs:83]: Please use expm1

**Comment by infinity0** [comment]
Thanks for the quick review! I've changed to using `expm1` and `log1p` as you said, and added some tests.

Since lognormal generates very large values in `quantile` I had to tweak `complQuantileCheck` to make it pass, I figured since the relative error was OK it was worth keeping the test enabled, but you should take a look. Likewise I had to set `prec_quantile_CDF _ = (64,64)`, `48` was insufficient, I hope this is OK too.

I also changed `lognormalDistrE` to forbid too-large values of sigma (in practise, > about 20) because these cause the variance to be `> m_huge`.

File: Statistics/Distribution/Weibull.hs (hunk 1 above)
Comment (infinity0) [reply Statistics/Distribution/Weibull.hs]: Thanks. I wasn't sure if `expm1` would be *worse* for large `x` which we might have here, it's not clear from its documentation.

**Comment by infinity0** [comment]
I've added another commit that adds constructors for those distributions in terms of the mean and stddev, rather than their formal parameters. Lemme know if you'd rather have them as a separate PR, I figured it was OK here since they're quite small.

The weibull one is not very satisfactory but it was the best one I could find online, and I still don't understand where that -1.086 number comes from, the paper I cite just plucks it out of thin air with no explanation...

**Comment by infinity0** [comment]
Thanks! Yes it was good to run them, I missed them at first. Caught a few typos and corner cases. :)

**Comment by Shimuuar** [comment]
I finally got around to merge this PR. Thanks a lot and sorry for the wait

**Pull request merged**